### DISCUSSION TITLE
Add a <CR> at the end of a slime

### DISCUSSION THREAD

**Discussion by ubaldot** [jpalardy/vim-slime#301] [Q&A]
I am working on a `Windows 10` machine with the following `Vim`:

```
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Sep 21 2021 22:13:38)
MS-Windows 32-bit console version
Included patches: 1-3452
```

I am using `Vim` plus `iPython` that runs in a `Powershell` terminal as shown in picture

![image](https://user-images.githubusercontent.com/10845685/137715700-da8ba23e-8a5a-4811-915a-57268bd3f78d.png)

when run any slime command from the editor (for example when I run `<C-c><C-c>` or when I map something like `nmap <C-c><C-x> :SlimeSendCurrentLine<CR>`, see my `_vimrc` below) I would like that the slime automatically add a `<CR>` so that I don't have to switch to the terminal, then hit `Enter` and then switch back to the editor. 

How to do that?

For completeness, here is my `_vimrc`
```
set encoding=utf-8
set nu
set tabstop=4 softtabstop=4
set shiftwidth=4
set expandtab
set smartindent
set nobackup
set backspace=indent,eol,start
set nocompatible              " required
set clipboard=unnamed
set laststatus=2
set statusline=%f "tail of the filename
set incsearch " for displaying while searching
set smartcase
set hidden
set noswapfile

" Set terminal with 256 colors
set termguicolors
colorscheme desert
set nofoldenable
set foldmethod=syntax
set foldlevelstart=20

" Set fonts for gvim
if has("gui_win32")
    set guifont=Consolas:h11:cANSI
endif


" Buffer management
nnoremap <leader>b :ls<CR>:b
map <C-Left> :bnext<CR>
map <C-Right> :bprev<CR>

" Open terminal below all splits
" For Cygwin
" cabbrev bterm bo term
" For Windows
cabbrev bterm bo terminal powershell

" ALE completion stuff to be set before the plugin.
let g:ale_completion_enabled = 1
let g:ale_completion_autoimport = 1


" --- VUNDLE PLUGIN STUFF BEGIN --------
filetype off                  " required
" Vundle plugin manager
"set the runtime path to include Vundle and initialize
set rtp+=C:/Users/MySelf/vimfiles/bundle/Vundle.vim
 call vundle#begin('C:/Users/yt75534/vimfiles/bundle')
" let Vundle manage Vundle, required
" add all your plugins here (note older versions of Vundle
" used Bundle instead of Plugin)
Plugin 'gmarik/Vundle.vim'
Plugin 'tmhedberg/SimpylFold.vim'
Plugin 'dense-analysis/ale'
Plugin 'ctrlpvim/ctrlp'
Plugin 'nvie/vim-flake8'
Plugin 'davidhalter/jedi-vim'
Plugin 'jpalardy/vim-slime'
call vundle#end()            " required
filetype plugin indent on    " required for Vundle
" ---- VUNDLE PLUGIN STUFF END ----------

" ---- Plugin tweak ------
" SimylFold tweak
" Enable folding with the spacebar
nnoremap <space> za

" vim-slime fix paste issues in ipython
let g:slime_target = "vimterminal"
let g:slime_cell_delimiter = "# %%"
nmap <F6> :call slime#send_cell()<CR>:SlimeSend0 "\r"<CR>
nmap <C-c><C-x> :SlimeSendCurrentLine<CR>

" Set jedi-vim
" To use Anaconda python
let g:jedi#environment_path = 'C:\Users\MySelf\Anaconda3\python.exe'
" To avoid that it opens the doc while autocompleting
autocmd FileType python setlocal completeopt-=preview

" Ctrlp stuff
set wildignore+=*\\_out\\*,*\\Cantata\\*
" ctrlp search from current director
let g:ctrlp_cmd='CtrlP :pwd'
let g:ctrlp_max_files=0


"   ALE linter stuff
let g:ale_linters = {'c': ['clangtidy','clangd'],'python':['flake8']}
let g:ale_fixers = {
\   'c':['astyle','clangtidy','remove_trailing_lines', 'trim_whitespace'], 
\}
let g:ale_c_clangtidy_checks = ['*']
map <C-a><C-a> :ALEGoToDefinition<CR>
map <C-a><C-z> :ALEFindReferences<CR>

" Matlab syntax stuff. To use the code checker use :make
source $VIMRUNTIME/macros/matchit.vim
filetype indent on
autocmd BufEnter *.m    compiler mlint

syntax on
" When using iPython to avoid that shift space gives 32;2u
tnoremap <S-space> <space>

" To have nice colors during autocompletion
hi Pmenu guibg=lightgray guifg=black
hi PmenuSel guibg=darkgray guifg=gray


```

**Comment by jpalardy** [comment]
Hi @Barzi2001 

how is it different than https://github.com/jpalardy/vim-slime/issues/299 ?

**Comment by ubaldot** [comment]
It’s different in the sense that https://github.com/jpalardy/vim-slime/issues/299 added an extra `—-` that has been solved by removing the line 
`let g:slime_python_ipython = 1`

Furthermore, you suggested me the mapping 
`nmap <F6> :call slime#send_cell()<CR>:SlimeSend0 "\r"<CR>`

that works as expected: it sends the code cell from the editor to the terminal AND it executes it. 

With the other slime commands the text is “slimed” from the editor to the terminal, no `—-`  is added (cool!) but then the command is NOT executed (missing `<CR>`?). I have to switch to the terminal and manually hit `Enter` to execute the slimed text.

So, I assume that I should map the slime commands that I want to use and add a `<CR>` at the end of the mapping as you suggested me to do with this

`nmap <F6> :call slime#send_cell()<CR>:SlimeSend0 "\r"<CR>`

But I have no clue how. 
I tried things like

`nmap <C-c><C-x> :SlimeSendCurrentLine<CR>`

But it doesn’t work, i.e. the current line is slimed towards the terminal but then the `<CR>` won’t happen.
How to solve my problem?
I hope this clarify a bit :)

**Comment by ubaldot** [comment]
**UPDATE!**
I think I may have found a bug or it is just me that is not understanding anything. :)

I have jut noticed that the more correct way of using vim-slime when you have a split "vim editor - terminal running iPython" is to set
`let g:slime_python_ipython = 1`
By doing this, the command `%cpaste -q` is called, but the newline/carriage-return does not seem to work nice under `Windows 10`. 
That is, vim-slime seems to use `\<CR>` instead of `<CR>`.
I explain it better. 

Assume that I want to execute the following code that I pasted it in the clipboard

```
a=3
b=4
c = a+b
```

**Option 1:**
Type `%cpaste -q` in `iPython`, hit `Enter`, paste the piece of code, type `--`, hit `Enter`.
This is the output

![image](https://user-images.githubusercontent.com/10845685/137870932-e9d30e4b-127a-483c-992a-dbdea0ddd937.png)

Everything works as expected.

**Option 2**
Slime the code from the editor to `iPython` with e.g. `<C-c><C-c>`.
This is the output

![image](https://user-images.githubusercontent.com/10845685/137871317-199821fa-a417-4892-b267-5e419368f705.png)

As you can see, the line after `%cpaste -q` starts with `...:` which to me suggests that there is something going wrong with the `<CR>`, i.e. it seems that the command `\<CR>` instead of `<CR>` is executed after `%cpaste -q`. 

In the picture below you can see what happens when you hit `Enter` at the end of a line or if you type `\` and then you hit `Enter` at the end of a line. 

![image](https://user-images.githubusercontent.com/10845685/137872674-7e1b3dc4-033b-4458-8035-322f61c253ee.png)

**Reply by ubaldot** [reply]
**Yet another update!**

I am not 100% sure, but it seems that the problem relies on Windows and some not yet implemented functionalities in iPython.
Here is an excerpt from [iPython documentation:](https://ipython.org/ipython-doc/rel-0.9.1/html/interactive/reference.html#caution-for-windows-users) 

```
%cpaste:

Allows you to paste & execute a pre-formatted code block from clipboard

You must terminate the block with '--' (two minus-signs) alone on the
line. You can also provide your own sentinel with '%paste -s %%' ('%%'
is the new sentinel for this operation)

The block is dedented prior to execution to enable execution of method
definitions. '>' and '+' characters at the beginning of a line are
ignored, to allow pasting directly from e-mails or diff files. The
executed block is also assigned to variable named 'pasted_block' for
later editing with '%edit pasted_block'.

You can also pass a variable name as an argument, e.g. '%cpaste foo'.
This assigns the pasted block to variable 'foo' as string, without
dedenting or executing it.

Do not be alarmed by garbled output on Windows (it's a readline bug).
Just press enter and type -- (and press enter again) and the block
will be what was just pasted.

IPython statements (magics, shell escapes) are not supported (yet).
```

Inspection of `ftplugin/python.slime.vim` shows that `line 8` uses `\n` as escape character for `Enter`: 

`    return ["%cpaste -q\n", g:slime_dispatch_ipython_pause, a:text, "--\n"]`

I have tried to replace `\n` with `\r\n` and `\r` but the problem is still there.

**Comment by jpalardy** [comment]
🤔 

if combination of `\n` and `\r` won't work… I'm not sure what's next.

@Barzi2001 let me know if you find out more.

**Reply by ubaldot** [reply]
Apparently it is not working. Latest test has been done by using double `\\` but the problem persists.
I am already asking around, but perhaps this is worth to be mentioned on the issue tracker (provided that it is a bug for Windows systems)?

**Comment by ubaldot** [comment]
I found a solution [here](https://issueexplorer.com/issue/jpalardy/vim-slime/283) 
It is enough to implement the following changes in `ftplugin/python.slime.vim`:

**line 8**
```
-    return ["%cpaste -q,\n", g:slime_dispatch_ipython_pause, a:text, "--\n"]
+    return ["%cpaste -q", "\n", g:slime_dispatch_ipython_pause, a:text, "--", "\n"]
```

**line 16**
```
-   return substitute(dedented_lines, add_eol_pat, "\n", "g")
+   return [substitute(dedented_lines, add_eol_pat, "", "g"), "\n"]
```

EDIT: perhaps **line 15** shall be adjusted as well since `:SlimeSendCurrentLine` seems to send an extra `\n`

@jpalardy can these changes been implemented, please? :)
